cpp/src/copying/concatenate.cu

```diff
@@ -335,26 +336,82 @@ std::unique_ptr<column> concatenate_dispatch::operator()<cudf::struct_view>()
   return cudf::structs::detail::concatenate(views, stream, mr);
 }
 
+namespace {
+
+/**
+ * @brief Verifies that the sum of the sizes of all the columns to be concatenated
+ * will not exceed the max value of size_type, and verifies all column types match
+ *
+ * @param begin Beginning of range of columns to check
+ * @param end End of range of columns to check
+ *
+ * @throws cudf::logic_error if the total length of the concatenated columns would
+ * exceed the max value of size_type
+ *
+ * @throws cudf::logic_error if all of the input column types don't match
+ */
+template <typename ColIter>
+void bounds_and_type_check(ColIter begin, ColIter end)
+{
+  CUDF_EXPECTS(std::all_of(begin,
+                           end,
+                           [expected_type = (*begin).type()](auto const& c) {
+                             return c.type() == expected_type;
+                           }),
+               "Type mismatch in columns to concatenate.");
+
+  // total size of all concatenated rows
+  auto size_iter = thrust::make_transform_iterator(
+    begin, [](column_view const& c) { return static_cast<size_t>(c.size()); });
+  size_t const total_row_count =
+    std::accumulate(size_iter, size_iter + std::distance(begin, end), static_cast<size_t>(0));
```

Comment:
This can be simpler since `accumulate` can take a lambda...

```suggestion
  size_t const total_row_count =
    std::accumulate(begin, end, std::size_t{}, [](auto const& c) { return static_cast<size_t>(c.size()); });
```

Reply:
👍

Reply:
Done